### ISSUE TITLE
qwen3_omni_moe audio path: three stacked bugs for non-generate use (batched chunker slice, sample-domain feature mask -> 300GB allocs, ~80x audio token over-expansion)

### ISSUE THREAD

**Issue by FreazeMood** [Blaizzy/mlx-vlm#1651]
### Summary

Extracting hidden states (or any non-generate use) of `qwen3_omni_moe` with real-length audio (>~5s) fails or silently degrades through three stacked bugs in the audio path. Found while running last-token feature extraction over 16 kHz mono WAVs on `mlx-community/Qwen3-Omni-30B-A3B-Instruct-4bit` (mlx-vlm 0.6.5, M4 Max). Repro + working patches for all three below.

### Bug 1 — audio chunker assumes unbatched `(mel, T)` input

`models/qwen3_omni_moe/audio.py`, chunk-padding loop:

```python
padded_feature[i, :, :chunk_len] = input_features[:, start_idx:end_idx]
```

`input_features` arrives batched `(1, 128, T)` from the standard path, so the slice cuts the **mel axis** instead of time and the assignment broadcasts garbage (`(1,1,100,200)` vs `(1,128,100)` shape error). Fix that worked for us:

```python
if input_features.ndim == 3:
    input_features = input_features[0]
```

(or slice `input_features[0, :, start_idx:end_idx]`).

### Bug 2 — `feature_attention_mask` is sample-domain, consumed as mel-domain

`thinker.py::get_audio_features` computes `feature_lens = feature_attention_mask.sum(-1)`, but the mask produced by the processor path is over **raw audio samples** (e.g. 320,000 for 20 s), not mel frames (2,000). The tower then builds `320000/100 = 3200` chunks and attempts absurd allocations — we saw `metal::malloc` requests of **138–311 GB** for 20–60 s clips. The same function's unpad block (`mx.take` over the flattened features by that mask) corrupts the features for the same reason.

Workaround that works: bypass the mask entirely and pass `audio_feature_lengths` in mel frames (`mx.array([input_features.shape[-1]])`). A real fix should normalize the mask domain (`mask.sum(-1) // hop_length`) or document that `audio_feature_lengths` is required.

### Bug 3 — audio placeholder expansion mints ~80x too many tokens

With the template/processor path, a 60 s clip yields `input_ids` of **62,422 tokens** (expansion appears sample-derived), while the audio tower actually outputs **391 tokens per 30 s window**. Forward passes take ~170 s instead of ~0.5 s. Workaround: query the tower for its true output length, splice exactly that many `<|audio_pad|>` ids manually, and inject the audio features at those positions — 60 s clip end-to-end drops from 173 s to 0.6 s (~290x).

### Repro sketch

```python
model, proc = load("mlx-community/Qwen3-Omni-30B-A3B-Instruct-4bit")
text = apply_chat_template(proc, model.config, "You hear this sound.", num_audios=1)
inputs = prepare_inputs(proc, audio=["any_20s_16k_mono.wav"], prompts=[text])
model.get_input_embeddings(inputs["input_ids"],
    input_features=inputs["input_features"],
    feature_attention_mask=inputs["feature_attention_mask"])
# -> shape error (bug 1); after fixing: 138GB metal::malloc (bug 2);
# via processor path instead: input_ids ~62k tokens for 60s audio (bug 3)
```

Happy to send PRs for all three (the fixes above are running in production for us — batch extraction at ~1.1 clips/s). Found these building behavioral-prediction infrastructure that reads activations rather than generations, which is presumably why the generate-centric paths hadn't surfaced them.

**Comment by FreazeMood** [comment]
Update: there is a **fourth bug** in the same family — the **video** placeholder expansion is also broken for `qwen3_omni_moe`: `prepare_inputs` with `videos=[...]` produces `input_ids` containing the vision bos/eos markers (151667/151668) but **zero** `<|video_pad|>` (151656) tokens between them — the processor's `video_token_ids` is `[None]`, so the expansion count is never computed. Features are produced correctly (`pixel_values_videos`, `video_grid_thw`), but no positions exist to scatter them into.

Workaround (same pattern as the audio fix): compute the true count from the grid — `N_v = t*h*w // merge_size**2` — and splice that many `<|video_pad|>` ids between the vision markers manually. With that, combined **video + audio + text** single-stream extraction works end to end (~6.4 s for a 5,471-token forward on an M4 Max, 4-bit).

So the full tally for non-generate multimodal use of this port: audio chunker batch-dim slice, sample-domain feature mask, audio token over-expansion, and video token non-expansion. All four have working local patches; PR offer stands.

**Comment by Blaizzy** [comment]
Thanks @FreazeMood!

Fixed in main ✅
